### PR TITLE
Introducing new **catchAll parameter to allow generating links with u…

### LINKED ISSUE #539: Consider adding a feature for catchall that roundtrips / characters
Tracking issue for https://github.com/aspnet/Home/issues/2669#issuecomment-395226030

Questions: Will new segment type be dispatching only or will it also be added to OG routing?

### PR DESCRIPTION
**Pull request by kichalla**
…nencoded values

[Fixes #539] Consider adding a feature for catchall that roundtrips / characters

### CODE REVIEW

File: src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs [hunk 1]
```diff
@@ -27,9 +27,17 @@ public static RoutePatternParameterPart ParseRouteParameter(string parameter)
 
             var startIndex = 0;
             var endIndex = parameter.Length - 1;
+            var encodeValue = true;
 
             var parameterKind = RoutePatternParameterKind.Standard;
-            if (parameter[0] == '*')
+
+            if (parameter.StartsWith("**"))
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs]: Use overload with StringComparison.Ordinal

**Review by JamesNK — APPROVED** [review]
Looks good to me but wait for Ryan to take a look before merging.

FYI I created https://github.com/aspnet/Routing/issues/720 to track updating docs

File: src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs [hunk 2]
```diff
@@ -88,16 +93,24 @@ public bool Accept(string value)
...
+                }
+                else
+                {
+                    _urlEncoder.Encode(Writer, value);
+                }
             }
             else
             {
-                _urlEncoder.Encode(Writer, value);
+                _uri.Append(value);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs]: We still need to encode characters other than `/`

File: src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParameterPart.cs [hunk 3]
```diff
@@ -27,13 +37,19 @@ public sealed class RoutePatternParameterPart : RoutePatternPart
...
 
         /// <summary>
         /// Gets the list of constraints associated with this parameter.
         /// </summary>
         public IReadOnlyList<RoutePatternConstraintReference> Constraints { get; }
 
+        /// <summary>
+        /// Gets the value indicating if current parameter's value should be encoded.
+        /// </summary>
+        public bool EncodeValue { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParameterPart.cs]: Should we call this something like `EncodeSlashes`?

File: test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs [hunk 4]
```diff
@@ -98,6 +98,40 @@ public void GetLink_MultipleEndpoints_Success2()
...
+            var linkGenerator = CreateLinkGenerator(endpoint);
+            var httpContext = CreateHttpContext(ambientValues: new { controller = "Home", action = "Index" });
+
+            // Act
+            var link = linkGenerator.GetLink(httpContext, new { path = routeValue });
+
+            // Assert
+            Assert.Equal(expected, link);
+        }
+
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: Can you add some cases where there are characters other than `/` that need to be url-encoded?

**Review by kichalla — COMMENTED** [review]
🆙 📅

File: src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs (hunk 2 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs]: I'm going to guess there is no built in way to exclude encoding character and we don't want to write our own encode method or Barry will smite us.

Split value by `/`, encode each part to the writer, separated by `/`?

File: src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs [hunk 5]
```diff
@@ -197,6 +203,38 @@ public override string ToString()
...
+            EncodeValue(value, startIndex: 0, characterCount: value.Length, encodeSlashes);
+        }
+
+        private void EncodeValue(string value, int startIndex, int characterCount, bool encodeSlashes)
+        {
+            // Just encode everything if its ok to encode slashes
+            if (!encodeSlashes && value.IndexOf('/', startIndex, characterCount) > 0)
+            {
+                var subValue = value.Substring(startIndex, characterCount);
+                var parts = subValue.Split('/');
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs]: We can avoid allocating with the `Substring` and `Split` with a loop using `value.IndexOf('/', lastSlashIndex, remainingCharacterCount)` to find the next slash and then writing that set of characters using `_urlEncoder.Encode(Writer, value, lastSlashIndex + 1, nextSlashIndex);` + a find call to Encode with the trailing chars.

File: test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs (hunk 4 above)
Comment (JamesNK) [reply test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: Some extra tests:
* multiple slashes in a row, e.g. localhost/test//blah
* catchall ends with a trailing slash, e.g. localhost/text/thisiscatchallcontent/

Comment (kichalla) [reply test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: yeah, i have those scenarios covered here actually

File: src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs (hunk 5 above)
Comment (kichalla) [reply src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs]: Good idea 👍

**Review by kichalla — COMMENTED** [review]
🆙 📅

File: src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs [hunk 6]
```diff
@@ -197,6 +203,44 @@ public override string ToString()
...
+            // Just encode everything if its ok to encode slashes
+            if (encodeSlashes)
+            {
+                _urlEncoder.Encode(Writer, value, startIndex, characterCount);
+            }
+            else
+            {
+                var length = startIndex + characterCount;
+                var lastStartIndex = startIndex;
+                for (var i = startIndex; i < length; i++)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs]: nit: I would expect this loop to look more like: https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing/Matching/FastPathTokenizer.cs#L29

# CHANNEL YOUR INNER C PROGRAMMER

File: src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs [hunk 7]
```diff
@@ -27,9 +27,17 @@ public static RoutePatternParameterPart ParseRouteParameter(string parameter)
...
             var startIndex = 0;
             var endIndex = parameter.Length - 1;
+            var encodeValue = true;
 
             var parameterKind = RoutePatternParameterKind.Standard;
-            if (parameter[0] == '*')
+
+            if (parameter.StartsWith("**", StringComparison.Ordinal))
+            {
+                encodeValue = false;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs]: rename to `encodeSlashes` to match

**Review by rynowak — APPROVED** [review]

**Review by JamesNK — APPROVED** [review]
:shipit:

**Pull request merged**